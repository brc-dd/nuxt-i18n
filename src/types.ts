import type { STRATEGIES } from './constants'

export type Strategies = typeof STRATEGIES[keyof typeof STRATEGIES]

export type CustomRoutePages = Record<string, Record<string, string>>

export interface LocaleInfo {
  code: string
  path: string
  file: string
}
