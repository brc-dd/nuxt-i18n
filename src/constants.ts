export const DEFAULT_LOCALE = ''
export const DEFAULT_TRAILING_SLASH = false
export const DEFAULT_ROUTES_NAME_SEPARATOR = '___'
