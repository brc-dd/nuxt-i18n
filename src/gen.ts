import { toCode } from './utils'
import type { ModuleOptions } from './module'

export function generateLoaderOptions(options: ModuleOptions) {
  const genCode = Object.entries(options)
    .map(([rootKey, rootValue]) => {
      return `export const ${rootKey} = ${toCode(rootValue)}`
    })
    .join('\n')

  return genCode
}
