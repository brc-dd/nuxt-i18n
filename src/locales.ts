import { parse } from 'pathe'
import { resolveFiles } from '@nuxt/kit'
import type { LocaleInfo } from './types'

export async function resolveLocales(path: string): Promise<LocaleInfo[]> {
  const files = await resolveFiles(path, '**/*{json,json5,yaml,yml}')
  return files.map((file) => {
    const parsed = parse(file)
    return {
      path: file,
      file: parsed.base,
      code: parsed.name,
    }
  })
}
