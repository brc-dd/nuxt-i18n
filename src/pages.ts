import { extendPages } from '@nuxt/kit'
import { localizeRoutes } from 'vue-i18n-routing'
import type { Nuxt, NuxtPage } from '@nuxt/schema'
import type { ComputedRouteOptions, RouteOptionsResolver } from 'vue-i18n-routing'
import type { ModuleOptions } from './module'

export function setupPages(
  options: Required<ModuleOptions>,
  nuxt: Nuxt,
) {
  let includeUprefixedFallback = nuxt.options.target === 'static'
  nuxt.hook('generate:before', () => {
    includeUprefixedFallback = true
  })

  extendPages((pages) => {
    const localizedPages = localizeRoutes(pages, {
      ...options,
      includeUprefixedFallback,
      optionsResolver: getRouteOptionsResolver(options),
    })
    pages.splice(0, pages.length)
    pages.unshift(...(localizedPages as NuxtPage[]))

    for (const [key, value] of Object.entries(options.routeOverrides)) {
      const page = pages.find(({ path }) => path === key)
      if (page)
        page.path = value
    }

    // if (nuxt.options.dev)
    //   logger.info('Localized pages:', pages)
  })
}

function getRouteOptionsResolver(moduleOptions: Required<ModuleOptions>): RouteOptionsResolver {
  return function (route, localeCodes): ComputedRouteOptions | null {
    const options: ComputedRouteOptions = {
      locales: localeCodes,
      paths: {},
    }

    // Set custom localized route paths
    if (Object.keys(moduleOptions.pages).length > 0) {
      for (const locale of options.locales) {
        const customPath = moduleOptions.pages[route.path]?.[locale]
        if (customPath)
          options.paths[locale] = customPath
      }
    }

    return options
  }
}
