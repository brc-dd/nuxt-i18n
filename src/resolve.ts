import type { NuxtPage } from '@nuxt/schema'
import {
  DEFAULT_LOCALE,
  DEFAULT_LOCALE_ROUTE_NAME_SUFFIX,
  DEFAULT_ROUTES_NAME_SEPARATOR,
  DEFAULT_STRATEGY,
  DEFAULT_TRAILING_SLASH,
} from './constants'
import type { Strategies } from './types'

export interface ComputedRouteOptions {
  locales: readonly string[]
  paths: Record<string, string>
}

export type RouteOptionsResolver = (
  route: NuxtPage,
  localeCodes: string[]
) => ComputedRouteOptions | null

export interface LocalizeRoutesPrefixableOptions {
  currentLocale: string
  defaultLocale: string
  strategy: Strategies
  isChild: boolean
  path: string
}

export type LocalizeRoutesPrefixable = (
  options: LocalizeRoutesPrefixableOptions
) => boolean

export interface I18nRoutingLocalizationOptions {
  /**
   * The app's default locale
   *
   * @default ''
   */
  defaultLocale?: string
  /**
   * List of locales supported by the app
   *
   * @default []
   */
  locales?: string[]
  /**
   * Routes strategy
   *
   * @remarks
   * Can be set to one of the following:
   *
   * - `no_prefix`: routes won't have a locale prefix
   * - `prefix_except_default`: locale prefix added for every locale except default
   * - `prefix`: locale prefix added for every locale
   * - `prefix_and_default`: locale prefix added for every locale and default
   *
   * @default 'prefix_except_default'
   */
  strategy?: Strategies
  /**
   * Whether to use trailing slash
   *
   * @default false
   */
  trailingSlash?: boolean
  /**
   * Internal separator used for generated route names for each locale
   *
   * @default '___'
   */
  routesNameSeparator?: string
  /**
   * Internal suffix added to generated route names for default locale
   *
   * @default 'default'
   */
  defaultLocaleRouteNameSuffix?: string
  /**
   * Whether to prefix the localize route path with the locale or not
   *
   * @default {@link DefaultLocalizeRoutesPrefixable}
   */
  localizeRoutesPrefixable?: LocalizeRoutesPrefixable
  /**
   * Whether to include uprefixed fallback route
   *
   * @default false
   */
  includeUprefixedFallback?: boolean
  /**
   * Resolver for route localizing options
   *
   * @default undefined
   */
  optionsResolver?: RouteOptionsResolver
}

function adjustRoutePathForTrailingSlash(
  pagePath: string,
  trailingSlash: boolean,
  isChildWithRelativePath: boolean,
) {
  return (
    pagePath.replace(/\/+$/, '') + (trailingSlash ? '/' : '')
    || (isChildWithRelativePath ? '' : '/')
  )
}

function prefixable(options: LocalizeRoutesPrefixableOptions): boolean {
  const { currentLocale, defaultLocale, strategy, isChild, path } = options

  const isDefaultLocale = currentLocale === defaultLocale
  const isChildWithRelativePath = isChild && !path.startsWith('/')

  // No need to add prefix if child's path is relative
  return (
    !isChildWithRelativePath
    // Skip default locale if strategy is `prefix_except_default`
    && !(isDefaultLocale && strategy === 'prefix_except_default')
  )
}

export const DefaultLocalizeRoutesPrefixable = prefixable

/**
 * Localize all routes with given locales
 *
 * @remarks
 * Based on [@intlify/routing](https://github.com/intlify/routing), licensed under MIT
 */
export function localizeRoutes(
  routes: NuxtPage[],
  {
    defaultLocale = DEFAULT_LOCALE,
    strategy = DEFAULT_STRATEGY as Strategies,
    trailingSlash = DEFAULT_TRAILING_SLASH,
    routesNameSeparator = DEFAULT_ROUTES_NAME_SEPARATOR,
    defaultLocaleRouteNameSuffix = DEFAULT_LOCALE_ROUTE_NAME_SUFFIX,
    includeUprefixedFallback = false,
    optionsResolver = undefined,
    localizeRoutesPrefixable = DefaultLocalizeRoutesPrefixable,
    locales = [],
  }: I18nRoutingLocalizationOptions = {},
): NuxtPage[] {
  if (strategy === 'no_prefix')
    return routes

  function makeLocalizedRoutes(
    route: NuxtPage,
    allowedLocaleCodes: string[],
    isChild = false,
    isExtraPageTree = false,
  ): NuxtPage[] {
    // Skip route localization
    if (route.redirect && !route.file)
      return [route]

    // Resolve with route (page) options
    let routeOptions: ComputedRouteOptions | null = null
    if (optionsResolver != null) {
      routeOptions = optionsResolver(route, allowedLocaleCodes)
      if (routeOptions == null)
        return [route]
    }

    // Component specific options
    const componentOptions: ComputedRouteOptions = {
      locales,
      paths: {},
    }

    if (routeOptions != null)
      Object.assign(componentOptions, routeOptions)

    Object.assign(componentOptions, { locales: allowedLocaleCodes })

    // Double check locales to remove any locales not found in `pageOptions`
    // This is there to prevent children routes being localized even though they are disabled in the configuration
    if (
      componentOptions.locales.length > 0
      && routeOptions
      && routeOptions.locales != null
      && routeOptions.locales.length > 0
    ) {
      componentOptions.locales = componentOptions.locales.filter(
        locale => routeOptions.locales.includes(locale),
      )
    }

    return componentOptions.locales.reduce<NuxtPage[]>((_routes, locale) => {
      const { name } = route
      let { path } = route
      const localizedRoute = { ...route }

      // Make localized page name
      if (name)
        localizedRoute.name = `${name}${routesNameSeparator}${locale}`

      // Generate localized children routes
      if (route.children) {
        localizedRoute.children = route.children.reduce<NonNullable<NuxtPage['children']>>(
          (children, child) => [
            ...children,
            ...makeLocalizedRoutes(child, [locale], true, isExtraPageTree),
          ],
          [],
        )
      }

      // Get custom path if any
      if (componentOptions.paths && componentOptions.paths[locale])
        path = componentOptions.paths[locale]

      // For `prefix_and_default` strategy and default locale:
      // - if it's a parent page, add it with default locale suffix added (no suffix if page has children)
      // - if it's a child page of that extra parent page, append default suffix to it
      const isDefaultLocale = locale === defaultLocale
      if (isDefaultLocale && strategy === 'prefix_and_default') {
        if (!isChild) {
          const defaultRoute = { ...localizedRoute, path }

          if (name)
            defaultRoute.name = `${localizedRoute.name}${routesNameSeparator}${defaultLocaleRouteNameSuffix}`

          if (route.children) {
            // Recreate child routes with default suffix added
            defaultRoute.children = []
            for (const childRoute of route.children) {
              // `isExtraPageTree` argument is true to indicate that this is extra route added for `prefix_and_default` strategy
              defaultRoute.children.push(
                ...makeLocalizedRoutes(childRoute, [locale], true, true),
              )
            }
          }

          _routes.push(defaultRoute)
        }
        else if (isChild && isExtraPageTree && name) {
          localizedRoute.name += `${routesNameSeparator}${defaultLocaleRouteNameSuffix}`
        }
      }

      const isChildWithRelativePath = isChild && !path.startsWith('/')

      // Add route prefix
      const shouldAddPrefix = localizeRoutesPrefixable({
        isChild,
        path,
        currentLocale: locale,
        defaultLocale,
        strategy,
      })
      if (shouldAddPrefix)
        path = `/${locale}${path}`

      if (path) {
        path = adjustRoutePathForTrailingSlash(
          path,
          trailingSlash,
          isChildWithRelativePath,
        )
      }

      if (shouldAddPrefix && isDefaultLocale && strategy === 'prefix' && includeUprefixedFallback)
        _routes.push({ ...route })

      localizedRoute.path = path
      _routes.push(localizedRoute)

      return _routes
    }, [])
  }

  return routes.reduce<NuxtPage[]>(
    (localized, route) => [
      ...localized,
      ...makeLocalizedRoutes(route, locales || []),
    ],
    [],
  )
}
