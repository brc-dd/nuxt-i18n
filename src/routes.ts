import { type NuxtPage } from '@nuxt/schema'
import { adjustRoutePathForTrailingSlash, getLocalizedRouteName } from './utils'

export interface ComputedRouteOptions {
  locales: readonly string[]
  paths: Record<string, string>
}

export type RouteOptionsResolver = (
  route: NuxtPage,
  localeCodes: string[]
) => ComputedRouteOptions | null

export interface LocalizeRoutesPrefixableOptions {
  currentLocale: string
  defaultLocale: string
  isChild: boolean
  path: string
}

export type LocalizeRoutesPrefixable = (
  options: LocalizeRoutesPrefixableOptions
) => boolean

export interface I18nRoutingLocalizationOptions {
  /**
   * The app's default locale
   *
   * @default ''
   */
  defaultLocale?: string
  /**
   * List of locales supported by the app
   *
   * @default []
   */
  locales?: string[]
  /**
   * Whether to use trailing slash
   *
   * @default false
   */
  trailingSlash?: boolean
  /**
   * Whether to prefix the localize route path with the locale or not
   *
   * @default {@link DefaultLocalizeRoutesPrefixable}
   */
  localizeRoutesPrefixable?: LocalizeRoutesPrefixable
  /**
   * Whether to include uprefixed fallback route
   *
   * @default false
   */
  includeUprefixedFallback?: boolean
  /**
   * Resolver for route localizing options
   *
   * @default undefined
   */
  optionsResolver?: RouteOptionsResolver
}

function prefixable(options: LocalizeRoutesPrefixableOptions): boolean {
  const { currentLocale, defaultLocale, isChild, path } = options

  const isDefaultLocale = currentLocale === defaultLocale
  const isChildWithRelativePath = isChild && !path.startsWith('/')

  // No need to add prefix if child's path is relative and current locale is default
  return !isChildWithRelativePath && !isDefaultLocale
}

export const DefaultLocalizeRoutesPrefixable = prefixable

/**
 * Localize all routes with given locales
 *
 * @remarks
 * Based on [@intlify/routing](https://github.com/intlify/routing), licensed under MIT
 */
export function localizeRoutes(
  routes: NuxtPage[],
  {
    defaultLocale = '',
    trailingSlash = false,
    optionsResolver = undefined,
    localizeRoutesPrefixable = DefaultLocalizeRoutesPrefixable,
    locales = []
  }: I18nRoutingLocalizationOptions = {}
): NuxtPage[] {
  function makeLocalizedRoutes(
    route: NuxtPage,
    allowedLocaleCodes: string[],
    isChild = false,
    isExtraPageTree = false
  ): NuxtPage[] {
    // Skip route localization
    if (route.redirect && !route.file) { return [route] }

    // Resolve with route (page) options
    let routeOptions: ComputedRouteOptions | null = null
    if (optionsResolver != null) {
      routeOptions = optionsResolver(route, allowedLocaleCodes)
      if (routeOptions == null) { return [route] }
    }

    // Component specific options
    const componentOptions: ComputedRouteOptions = {
      locales,
      paths: {}
    }

    if (routeOptions != null) {
      Object.assign(componentOptions, routeOptions)
    }

    Object.assign(componentOptions, { locales: allowedLocaleCodes })

    // Double check locales to remove any locales not found in `pageOptions`
    // This is there to prevent children routes being localized even though they are disabled in the configuration
    if (
      componentOptions.locales.length > 0
      && routeOptions
      && routeOptions.locales != null
      && routeOptions.locales.length > 0
    ) {
      componentOptions.locales = componentOptions.locales.filter(
        (locale) => routeOptions!.locales.includes(locale)
      )
    }

    return componentOptions.locales.reduce<NuxtPage[]>((_routes, locale) => {
      const { name } = route
      let { path } = route
      const localizedRoute = { ...route }

      // Make localized page name
      if (name) {
        localizedRoute.name = getLocalizedRouteName(name, locale)
      }

      // Generate localized children routes
      if (route.children) {
        localizedRoute.children = route.children.reduce<NonNullable<NuxtPage['children']>>(
          (children, child) => [
            ...children,
            ...makeLocalizedRoutes(child, [locale], true, isExtraPageTree)
          ],
          []
        )
      }

      // Get custom path if any
      if (componentOptions.paths && componentOptions.paths[locale]) {
        path = componentOptions.paths[locale]
      }

      const isChildWithRelativePath = isChild && !path.startsWith('/')

      // Add route prefix
      const shouldAddPrefix = localizeRoutesPrefixable({
        isChild,
        path,
        currentLocale: locale,
        defaultLocale
      })
      if (shouldAddPrefix) {
        path = `/${locale}${path}`
      }

      if (path) {
        path = adjustRoutePathForTrailingSlash(
          path,
          trailingSlash,
          isChildWithRelativePath
        )
      }

      localizedRoute.path = path
      _routes.push(localizedRoute)

      return _routes
    }, [])
  }

  return routes.reduce<NuxtPage[]>(
    (localized, route) => [
      ...localized,
      ...makeLocalizedRoutes(route, locales || [])
    ],
    []
  )
}
