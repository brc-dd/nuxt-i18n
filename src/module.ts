import { resolve as pathResolve } from 'pathe'
import { addImportsDir, addPluginTemplate, addTemplate, createResolver, defineNuxtModule } from '@nuxt/kit'
import { genDynamicImport, genImport, genSafeVariableName } from 'knitwork'
import type { I18nRoutingOptions } from 'vue-i18n-routing'
import { setupPages } from './pages'
import { resolveLocales } from './locales'
import { toCode } from './utils'
import type { LocaleInfo } from './types'

export type ModuleOptions = {
  /**
   * List of locales supported by your app
   *
   * @remarks
   * Intended to be an array of string codes, e.g. `['en', 'fr']`
   *
   * @default []
   */
  locales?: string[]

  /**
   * The app's default locale
   *
   * @remarks
   * It's recommended to set this to some locale regardless of chosen strategy, as it will be used as a fallback locale when navigating to a non-existent route
   *
   * @default 'en'
   */
  defaultLocale?: string

  /**
   * Directory where your locale files are stored
   *
   * @remarks
   * Expected to be a relative path from the project root
   *
   * @default 'locales'
   */
  langDir?: string

  /**
   * Whether to enable locale auto-importing
   *
   * @remarks
   * When enabled, the module will automatically import all locale files from the `langDir` directory
   *
   * @default false
   */
  langImports?: boolean

  /**
   * Whether to lazily load locale messages in the client
   *
   * @remarks
   * If enabled, locale messages will be loaded on demand when the user navigates to a route with a different locale
   *
   * This has no effect if `strategy` is set to `no_prefix` or the `langImports` option is disabled
   *
   * @default true
   */
  lazy?: boolean

  /**
   * The app's default messages
   *
   * @remarks
   * Can be left empty if auto-importing of locales is enabled
   *
   * @default {}
   */
  messages?: Record<string, any>

  /**
   * Custom route overrides for the generated routes
   *
   * @example
   * routeOverrides: {
   *   // Set `en` (default locale) index page as the app's root page
   *   '/en': '/',
   * }
   *
   * @default {}
   */
  routeOverrides?: Record<string, string>

  /**
   * Localize route paths for a given locale
   *
   * @default {}
   */
  // pages?: CustomRoutePages
} & Pick<I18nRoutingOptions, 'strategy'>

export default defineNuxtModule<ModuleOptions>({
  meta: {
    name: '@leanera/nuxt-i18n',
    configKey: 'i18n',
    compatibility: {
      nuxt: '^3',
      bridge: false,
    },
  },
  defaults: {
    defaultLocale: 'en',
    locales: [],
    langDir: 'locales',
    langImports: false,
    messages: {},
    strategy: 'no_prefix',
    // pages: {},
    routeOverrides: {},
    lazy: true,
  },
  async setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url)
    const langPath = options.langImports && options.langDir ? pathResolve(nuxt.options.srcDir, options.langDir!) : undefined
    const localeInfo = langPath ? await resolveLocales(langPath) : []

    const syncLocaleFiles = new Set<LocaleInfo>()
    const asyncLocaleFiles = new Set<LocaleInfo>()

    if (langPath) {
      // Synchronously import locale messages for the default locale
      if (options.defaultLocale) {
        const localeObject = localeInfo.find(({ code }) => code === options.defaultLocale)
        if (localeObject)
          syncLocaleFiles.add(localeObject)
      }

      // Asynchronously import locale messages for the other locales
      for (const locale of localeInfo) {
        if (!syncLocaleFiles.has(locale) && !asyncLocaleFiles.has(locale))
          (options.lazy ? asyncLocaleFiles : syncLocaleFiles).add(locale)
      }
    }

    // Transpile runtime
    nuxt.options.build.transpile.push(resolve('runtime'))

    // Setup internationalized Nuxt pages
    if (options.strategy !== 'no_prefix' && options.locales?.length)
      setupPages(options as Required<ModuleOptions>, nuxt)

    // Add i18n plugin
    addPluginTemplate(resolve('runtime/plugin'))

    // Add i18n composables
    addImportsDir(resolve('runtime/composables'))

    // Load options template
    addTemplate({
      filename: 'i18n.options.mjs',
      getContents() {
        return `
${[...syncLocaleFiles]
  .map(({ code, path }) => genImport(path, genSafeVariableName(`locale_${code}`)))
  .join('\n')}

export const options = {${Object.entries(options)
  .map(([key, value]) => `${key}: ${toCode(value)}`)
  .join(',')}};
export const localeMessages = {
${[...syncLocaleFiles]
  .map(({ code }) => `  ${toCode(code)}: () => Promise.resolve(${genSafeVariableName(`locale_${code}`)}),`)
  .join('\n')}
${[...asyncLocaleFiles]
  .map(({ code, path }) => `  ${toCode(code)}: ${genDynamicImport(path)},`)
  .join('\n')}
};
`.trimStart()
      },
    })

    addTemplate({
      filename: 'i18n.options.d.ts',
      getContents() {
        return `
${genImport(resolve('module'), ['ModuleOptions'])}
export declare const options: Required<ModuleOptions>;
export declare const localeMessages: Record<string, () => Promise<Record<string, any>>>;
`.trimStart()
      },
    })
  },
})
