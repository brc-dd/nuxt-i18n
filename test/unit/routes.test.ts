import { type NuxtPage } from '@nuxt/schema'
import { describe, expect, it } from 'vitest'
import { localizeRoutes } from '../../src/routes'

// Adapted from: https://github.com/intlify/routing/blob/166e3c533ee47b40bc08e3af001bef33dcf975ed/packages/vue-i18n-routing/src/__test__/resolve.test.ts
// Credit: Kazuya Kawaguchi, @intlify
// License: MIT
describe('localizeRoutes', () => {
  describe('basic', () => {
    it('should be localized routing', () => {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        locales: localeCodes
      })

      expect(localizedRoutes).toMatchSnapshot()
      expect(localizedRoutes.length).to.equal(4)

      for (const locale of localeCodes) {
        for (const route of routes) {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}___${locale}`
          })
        }
      }
    })
  })

  describe('has children', () => {
    it('should be localized routing', () => {
      const routes: NuxtPage[] = [
        {
          path: '/user/:id',
          name: 'user',
          children: [
            {
              path: 'profile',
              name: 'user-profile'
            },
            {
              path: 'posts',
              name: 'user-posts'
            }
          ]
        }
      ]
      const children = routes[0]!.children!

      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        locales: localeCodes
      })

      expect(localizedRoutes).toMatchSnapshot()
      expect(localizedRoutes.length).to.equal(2)

      for (const locale of localeCodes) {
        for (const route of routes) {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}___${locale}`,
            children: children.map((child) => ({
              path: child.path,
              name: `${child.name}___${locale}`
            }))
          })
        }
      }
    })
  })

  describe('trailing slash', () => {
    it('should be localized routing', () => {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        locales: localeCodes,
        trailingSlash: true
      })

      expect(localizedRoutes).toMatchSnapshot()
      expect(localizedRoutes.length).to.equal(4)

      for (const locale of localeCodes) {
        for (const route of routes) {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}/`,
            name: `${route.name}___${locale}`
          })
        }
      }
    })
  })

  describe('strategy: "prefix_and_default"', () => {
    it('should be localized routing', () => {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        defaultLocale: 'en',
        locales: localeCodes
      })

      expect(localizedRoutes).toMatchSnapshot()
    })
  })

  describe('route option resolver: routing disable', () => {
    it('should be disabled routing', () => {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        locales: localeCodes,
        optionsResolver: () => null
      })

      expect(localizedRoutes).toMatchSnapshot()
    })
  })
})
