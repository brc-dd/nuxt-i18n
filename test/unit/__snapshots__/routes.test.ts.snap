// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`localizeRoutes > basic > should be localized routing 1`] = `
[
  {
    "name": "home___en",
    "path": "/en",
  },
  {
    "name": "home___ja",
    "path": "/ja",
  },
  {
    "name": "about___en",
    "path": "/en/about",
  },
  {
    "name": "about___ja",
    "path": "/ja/about",
  },
]
`;

exports[`localizeRoutes > has children > should be localized routing 1`] = `
[
  {
    "children": [
      {
        "name": "user-profile___en",
        "path": "profile",
      },
      {
        "name": "user-posts___en",
        "path": "posts",
      },
    ],
    "name": "user___en",
    "path": "/en/user/:id",
  },
  {
    "children": [
      {
        "name": "user-profile___ja",
        "path": "profile",
      },
      {
        "name": "user-posts___ja",
        "path": "posts",
      },
    ],
    "name": "user___ja",
    "path": "/ja/user/:id",
  },
]
`;

exports[`localizeRoutes > route option resolver: routing disable > should be disabled routing 1`] = `
[
  {
    "name": "home",
    "path": "/",
  },
  {
    "name": "about",
    "path": "/about",
  },
]
`;

exports[`localizeRoutes > strategy: "prefix_and_default" > should be localized routing 1`] = `
[
  {
    "name": "home___en",
    "path": "/",
  },
  {
    "name": "home___ja",
    "path": "/ja",
  },
  {
    "name": "about___en",
    "path": "/about",
  },
  {
    "name": "about___ja",
    "path": "/ja/about",
  },
]
`;

exports[`localizeRoutes > trailing slash > should be localized routing 1`] = `
[
  {
    "name": "home___en",
    "path": "/en/",
  },
  {
    "name": "home___ja",
    "path": "/ja/",
  },
  {
    "name": "about___en",
    "path": "/en/about/",
  },
  {
    "name": "about___ja",
    "path": "/ja/about/",
  },
]
`;
